src/kernel/cuda/spmm.cuh

```diff
@@ -0,0 +1,272 @@
+        arg_u[ty * out_len + tx] = local_argu;
+      if (ReduceOp::require_arg && BinaryOp::use_rhs)
+        arg_e[ty * out_len + tx] = local_arge;
+      tx += stride_x;
+    }
+    ty += stride_y;
+  }
+}
+
+#define BCAST_IDX_CTX_SWITCH(BCAST, EDGE_MAP, CTX, LHS_OFF, RHS_OFF, ...) do { \
+  const BcastOff &info = (BCAST);                                              \
+  if (!info.use_bcast) {                                                       \
+    constexpr bool UseBcast = false;                                           \
+    if ((EDGE_MAP)) {                                                          \
+      constexpr bool UseIdx = true;                                            \
+      { __VA_ARGS__ }                                                          \
+    } else {                                                                   \
+      constexpr bool UseIdx = false;                                           \
+      { __VA_ARGS__ }                                                          \
+    }                                                                          \
+  } else {                                                                     \
+    constexpr bool UseBcast = true;                                            \
+    DLContext ctx = (CTX);                                                     \
+    auto device = runtime::DeviceAPI::Get(ctx);                                \
+    (LHS_OFF) = static_cast<int64_t*>(                                         \
+      device->AllocWorkspace(ctx, sizeof(int64_t) * info.lhs_offset.size()));  \
+    CUDA_CALL(cudaMemcpy((LHS_OFF), &info.lhs_offset[0],                       \
+      sizeof(int64_t) * info.lhs_offset.size(), cudaMemcpyHostToDevice));      \
+    (RHS_OFF) = static_cast<int64_t*>(                                         \
+      device->AllocWorkspace(ctx, sizeof(int64_t) * info.rhs_offset.size()));  \
+    CUDA_CALL(cudaMemcpy((RHS_OFF), &info.rhs_offset[0],                       \
+      sizeof(int64_t) * info.rhs_offset.size(), cudaMemcpyHostToDevice));      \
+    if ((EDGE_MAP)) {                                                          \
+      constexpr bool UseIdx = true;                                            \
+      { __VA_ARGS__ }                                                          \
+    } else {                                                                   \
+      constexpr bool UseIdx = false;                                           \
+      { __VA_ARGS__ }                                                          \
+    }                                                                          \
+    device->FreeWorkspace(ctx, (LHS_OFF));                                     \
+    device->FreeWorkspace(ctx, (RHS_OFF));                                     \
+  }                                                                            \
+} while (0)                                                 
```

Comment:
Is this the same macro as the one in `sddmm.cuh`?

Reply:
Exactly

Reply:
Then pls merge them